drivers/display/display_ist3931.c

```diff
@@ -0,0 +1,318 @@
+	ist3931_write_bus_i2c(dev, &cmd_buf2, 1, 1);
+	return 0;
+};
+
+static inline int ist3931_driver_display_control(const struct device *dev)
+{
+	const struct ist3931_config *config = dev->config;
+	uint8_t cmd_buf1 = IST3931_CMD_Driver_Display_Control | config->shl << 3 |
+			   config->adc << 2 | config->eon << 1 | config->rev;
+	ist3931_write_bus_i2c(dev, &cmd_buf1, 1, 1);
+	return 0;
+};
+
+static inline int ist3931_driver_set_display_on(const struct device *dev)
+{
+	uint8_t cmd_buf1 = IST3931_CMD_Display_ON_OFF | 1;
+	ist3931_write_bus_i2c(dev, &cmd_buf1, 1, 1);
+	return 0;
+};
+
+static inline int ist3931_driver_set_com_pad_map(const struct device *dev)
+{
+	uint8_t cmd_buf[5] = {
+		IST3931_CMD_IST_COMMAND_ENTRY, IST3931_CMD_IST_COMMAND_ENTRY,
+		IST3931_CMD_IST_COMMAND_ENTRY, IST3931_CMD_IST_COMMAND_ENTRY,
+		IST3931_CMD_IST_COM_MAPPING,
+	};
+
+	uint8_t cmd_buf3 = IST3931_CMD_EXIT_ENTRY;
+	ist3931_write_bus_i2c(dev, &cmd_buf[0], 1, 5);
+	k_sleep(K_MSEC(IST3931_CMD_DELAY));
+	ist3931_write_bus_i2c(dev, &cmd_buf3, 1, 1);
+	return 0;
+};
+
+static inline int ist3931_driver_set_ay(const struct device *dev, uint16_t y)
+{
+	const struct ist3931_config *config = dev->config;
+	uint8_t cmd_buf1 = IST3931_CMD_Set_AY_ADD_LSB | ((config->y_offset + y) & 0x0F);
+	uint8_t cmd_buf2 = IST3931_CMD_Set_AY_ADD_MSB | ((config->y_offset + y) >> 4);
+	uint8_t cmd_buf[2] = {cmd_buf1, cmd_buf2};
+	ist3931_write_bus_i2c(dev, &cmd_buf[0], 1, 2);
+	return 0;
+};
+
+static inline int ist3931_driver_set_ax(const struct device *dev, uint8_t x)
+{
+	const struct ist3931_config *config = dev->config;
+	uint8_t cmd_buf1 = IST3931_CMD_Set_AX_ADD | (config->x_offset + x);
+	ist3931_write_bus_i2c(dev, &cmd_buf1, 1, 1);
+	return 0;
+};
+
+static int ist3931_init_device(const struct device *dev)
+{
+	const struct ist3931_config *config = dev->config;
+
+	// Reset IC
+	gpio_pin_set_dt(&config->reset_gpio, 1);
+	k_sleep(K_MSEC(IST3931_RESET_DELAY));
+	gpio_pin_set_dt(&config->reset_gpio, 0);
+	k_sleep(K_MSEC(IST3931_RESET_DELAY));
+	gpio_pin_set_dt(&config->reset_gpio, 1);
+	k_sleep(K_MSEC(IST3931_RESET_DELAY));
+	ist3931_set_power(dev);
+	ist3931_set_bias(dev);
+	ist3931_set_ct(dev);
+	ist3931_set_fr(dev);
+	ist3931_set_duty(dev);
+	ist3931_driver_display_control(dev);
+	ist3931_driver_set_display_on(dev);
+	ist3931_driver_set_com_pad_map(dev);
+	return 0;
+}
+
+static int ist3931_init(const struct device *dev)
+{
+	const struct ist3931_config *config = dev->config;
+	if (!ist3931_bus_ready(dev)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	} else {
+		LOG_INF("I2C device ready");
+	}
+	if (config->reset_gpio.port != NULL) {
+		if (!gpio_is_ready_dt(&config->reset_gpio)) {
+			LOG_ERR("Reset GPIO device not ready");
+			return -ENODEV;
+		}
+		if (gpio_pin_configure_dt(&config->reset_gpio, GPIO_OUTPUT_INACTIVE)) {
+			LOG_ERR("Couldn't configure reset pin");
+			return -EIO;
+		}
+
+		if (ist3931_init_device(dev)) {
```

Comment:
This feels incorrect- we only setup the device if a reset GPIO is present? If one was not present, what code would setup the display driver during boot?

Reply:
This display controller need a reset gpio to operate normally. Do I need to add some extra code to handle the miss of a reset gpio?

Reply:
No, instead you should make the `reset-gpios` property in the devicetree binding be `required: true`, and then you should check to make sure the `reset_gpio` is ready at boot (look at `gpio_is_ready_dt`). If the reset GPIO device isn't ready, the initialization function should return an error.

Reply:
Thanks, I will make the `reset-gpios` property be `required:true` in dts binding. The status of reset gpio is already check in line 204 `if (!gpio_is_ready_dt(&config->reset_gpio)) {
	LOG_ERR("Reset GPIO device not ready");
	return -ENODEV;
}`.

Reply:
This still needs to be addressed- `reset-gpios` is now `required:true`, but the `if (config->reset_gpio.port != NULL) {` block here wrapping this code should be dropped. We also should propagate the return codes from functions when they fail. I believe this code block should then look like so:

```
int ret;

if (!gpio_is_ready_dt(&config->reset_gpio)) {
	LOG_ERR("Reset GPIO device not ready");
	return -ENODEV;
}
ret = gpio_pin_configure_dt(&config->reset_gpio, GPIO_OUTPUT_INACTIVE);
if (ret) {
	LOG_ERR("Couldn't configure reset pin");
	return ret;
}

ret = ist3931_init_device(dev)
if (ret) {
	LOG_ERR("Failed to initialize device!");
	return ret;
}
```